SLEP006 - Add Metadata Routing to LogisticRegressionCV

OmarManzoor opened the pull request:
<!--
Thanks for contributing a pull request! Please ensure you have taken a look at
the contribution guidelines: https://github.com/scikit-learn/scikit-learn/blob/main/CONTRIBUTING.md
-->

#### Reference Issues/PRs
<!--
Example: Fixes #1234. See also #3456.
Please use keywords (e.g., Fixes) to create link to the issues or pull requests
you resolved, so that they will automatically be closed when your pull request
is merged. See https://github.com/blog/1506-closing-issues-via-pull-requests
-->
Towards: #22893

#### What does this implement/fix? Explain your changes.
- Added meta data routing to LogisticRegressionCV.
- Added add_self to specify that LogisticRegressionCV is also a consumer.
- Added routing for the cv and score methods

#### Any other comments?
- I might need some guidance on where to add further tests for this since it is not exactly a meta estimator.

<!--
Please be aware that we are a loose team of volunteers so patience is
necessary; assistance handling other issues is very welcome. We value
all user contributions, no matter how minor they are. If we are slow to
review, either the pull request needs some benchmarking, tinkering,
convincing, etc. or more likely the reviewers are simply busy. In either
case, we ask for your understanding during the review process.
For more information, see our FAQ on this topic:
http://scikit-learn.org/dev/faq.html#why-is-my-pull-request-not-getting-any-attention.

Thanks for contributing!
-->

OmarManzoor committed:
SLEP006 - Add Metadata Routing to LogisticRegressionCV

jnothman reviewed (commented):
Thanks for this @omar! Nice work! 
I agree it's hard to know what to test, and we should probably define expectations for this family of PRs.

I'd love to see a test, or an example, that uses a `cv=GroupKFold(n_splits=5)` or similar, which is a real motivating use case.

I can also see an error here: if the scorer requests `sample_weight` it won't receive it, and, unless I'm much mistaken, this will happen silently (i.e. there will be no error).

To fix this, we need to pass `sample_weight` to `process_routing` in `fit`. We would benefit from a test that the scorer receives `sample_weight` if requested.

OmarManzoor commented:
> > Thanks for this @omar! Nice work! I agree it's hard to know what to test, and we should probably define expectations for this family of PRs.
> > I'd love to see a test, or an example, that uses a `cv=GroupKFold(n_splits=5)` or similar, which is a real motivating use case.
> > I can also see an error here: if the scorer requests `sample_weight` it won't receive it, and, unless I'm much mistaken, this will happen silently (i.e. there will be no error).
> > To fix this, we need to pass `sample_weight` to `process_routing` in `fit`. We would benefit from a test that the scorer receives `sample_weight` if requested.
> 
> Thank you for the guidance!

I can't quite understand properly how to route the scoring parameters. Actually even if I pass sample weight here

`routed_params = process_routing(
            obj=self,
            method="fit",
            other_params=fit_params,
            sample_weight=sample_weight,
        )`
     
we still need someway to specify the child method of the scorer. However the available methods only contain score whereas any individual scorers themselves do not seem seem to define a score method. They instead have _score which is called through the __call__ method of the base scorer. Could you kindly guide me a bit here?

jnothman commented:
Scorers don't have a score method, but for the purpose of allowing the user to set routing requests, we call it 'score'. So when the scorer is applied, we need to pass it the parameters that are routed to it by the name 'score'.  Does that help, or is a code snippet more the right kind of support here?

OmarManzoor committed:
Adjust sample weight to scorer in fit method

OmarManzoor committed:
Fix the docstring error

adrinjalali commented on sklearn/linear_model/_logistic.py:
```diff
@@ -2016,15 +2028,54 @@ def score(self, X, y, sample_weight=None):
+        )
+
+        return scoring(
+            self, X, y, sample_weight=sample_weight, **routed_params.scorer.score
```
`routed_params.scorer.score` would include `sample_weight`, no need to pass it explicitly.

adrinjalali commented on sklearn/linear_model/_logistic.py:
```diff
@@ -2016,15 +2028,54 @@ def score(self, X, y, sample_weight=None):
+                scorer=self._get_scorer(),
+                method_mapping=MethodMapping()
+                .add(callee="score", caller="score")
+                .add(callee="score", caller="fit"),
```
fit is not passing extra params to scorer at the moment. I think that's missing

OmarManzoor replied:
Do you mean we should pass the params to the _log_reg_scoring_path?

adrinjalali replied:
yes.

adrinjalali reviewed (commented):
These PRs should also add the meta estimators here: https://github.com/scikit-learn/scikit-learn/blob/sample-props/sklearn/tests/test_metaestimators_metadata_routing.py

@jnothman anything else you think should be tested on all meta-estimators apart from what's there?

OmarManzoor commented:
> https://github.com/scikit-learn/scikit-learn/blob/sample-props/sklearn/tests/test_metaestimators_metadata_routing.py

How would I actually define the estimator_name for this case?

adrinjalali commented:
@OmarManzoor your link is to a file, I'm not sure what you're referring to :)

OmarManzoor commented:
> @OmarManzoor your link is to a file, I'm not sure what you're referring to :)

For example
`{
        "metaestimator": MultiOutputRegressor,
        "estimator_name": "estimator",
        "estimator": ConsumingRegressor,
        "X": X,
        "y": y_multi,
        "routing_methods": ["fit", "partial_fit"],
        "warns_on": {
            "fit": ["sample_weight", "metadata"],
            "partial_fit": ["sample_weight"],
        },
    },`

We need the estimator_name here. For meta estimators this should be just the estimator parameter name. In this case it is a little different.

OmarManzoor committed:
Send the extra scoring params to _log_reg_scoring_path

OmarManzoor committed:
Refactor

adrinjalali commented:
Oh I see what you mean. This meta-estimator doesn't actually have any sub-estimators the way others do.

So we kinda need to replicate the same kinds of tests, but for `scorer` and `cv` in this case. @BenjaminBossan might have more concrete ideas.

Ideally we'd have them in a way to use them for other similar *CV classes.

OmarManzoor commented:
> Oh I see what you mean. This meta-estimator doesn't actually have any sub-estimators the way others do.
> 
> So we kinda need to replicate the same kinds of tests, but for `scorer` and `cv` in this case. @BenjaminBossan might have more concrete ideas.
> 
> Ideally we'd have them in a way to use them for other similar *CV classes.

@adrinjalali Could you kindly clarify one thing. For scorers it seems like we won't raise any warnings or errors since when routing we actually search for the actual method specified, `return getattr(self, method)._route_params(params=params)`. In the case of scorers there is no "score" method so there is technically no set request. So when writing tests should we ignore both the warning and error tests?

adrinjalali commented:
For scorers we would raise an error if routing is not correct. For instance, if the scorer supports `sample_weight`, and no request is set for `score`, but `sample_weight` is passed, we raise.

We don't warn for them since the existing code doesn't do the routing, therefore all of it would be new functionality.

I'm not sure why you say the `test_setting_request_removes_warning_or_error` test is not beneficial. I'm happy to add more tests, would you like to write it and we review here?

OmarManzoor commented:
> For scorers we would raise an error if routing is not correct. For instance, if the scorer supports `sample_weight`, and no request is set for `score`, but `sample_weight` is passed, we raise.
> 
> We don't warn for them since the existing code doesn't do the routing, therefore all of it would be new functionality.
> 
> I'm not sure why you say the `test_setting_request_removes_warning_or_error` test is not beneficial. I'm happy to add more tests, would you like to write it and we review here?

Well if an error is raised then I think we can use the existing tests. However in that case I am not clear on how the error would be raised, since the underlying scorer does not have a "score" method whereas we try to check the signature of the score method since the "score" method is passed in the routing processing.

adrinjalali commented:
The scorer is the `score` method itself, and the meta-estimator is the one doing the routing, and that's where the error would occur if anything is wrong.

Instead of checking the signature, we should now check the routing of the scorer object. Scorers expose routing the same way as estimators do.

OmarManzoor commented:
> The scorer is the `score` method itself, and the meta-estimator is the one doing the routing, and that's where the error would occur if anything is wrong.
> 
> Instead of checking the signature, we should now check the routing of the scorer object. Scorers expose routing the same way as estimators do.

So I was trying to create a ConsumingScorer and use it within the existing tests structure in test_metaestimators_metadata_routing. But I seem to be doing it incorrectly. Would it be okay for me to push the code so that you can take a look and guide me?

adrinjalali commented:
Sure, happy to check.

OmarManzoor committed:
Attempt to add tests for scorers

OmarManzoor commented:
> Sure, happy to check.

Pushed the code. It does not raise an error for the function test_error_for_other_methods.

OmarManzoor committed:
Merge sample-props

OmarManzoor commented:
@adrinjalali Any feedback on this? Is it incorrect to test for the ConsumingScorer like this? Instead should we actually check the actual routing data and decide based on that?

adrinjalali commented:
I'm trying to figure out why the CI fails here. But I get a build not found error when I try to look. Could you please send an empty commit? Hoping to see what the issue here is.

OmarManzoor committed:
Empty Commit

OmarManzoor commented:
> I'm trying to figure out why the CI fails here. But I get a build not found error when I try to look. Could you please send an empty commit? Hoping to see what the issue here is.

This basically fails because we have a method test_error_for_other_methods which expects an error should be raised when we don't explicitly set the score request for sample weight, but we pass in sample weight anyways. The reason the scorer won't raise such an error is because we have no direct mapping to the score method of the underlying score since it only contains an _score method.

OmarManzoor committed:
Update logic in _route_params to set unrequested params for score

OmarManzoor committed:
Revert the latest change

OmarManzoor committed:
Added tests to check process routing for scorer in Logistic Regression CV fit

adrinjalali commented on sklearn/linear_model/_logistic.py:
```diff
@@ -1683,6 +1688,10 @@ def fit(self, X, y, sample_weight=None):
             If not provided, then each sample is given unit weight.
 
+        **fit_params : dict
+            Parameters to pass to the `fit` method of the underlying
+            classifier.
```
the underlying classifier wouldn't actually accept anything here. These parameters are only used to be passed to scorer and splitter, right?

adrinjalali commented on sklearn/linear_model/_logistic.py:
```diff
@@ -1798,6 +1814,11 @@ def fit(self, X, y, sample_weight=None):
         else:
             prefer = "processes"
 
+        # Remove the sample_weight parameter from score_params as it is
+        # explicitly being passed in the path_func.
+        score_params = routed_params.scorer.score
+        score_params.pop("sample_weight", None)
```
those sample weights are used for _fit_, while the score params passed are used to score. The `_log_reg_scoring_path` with the current implementation doesn't take `sample_weight` into account for scoring since it doesn't pass it to the score method if `sample_weight` is passed. Either here it needs to not be popped, or there it needs to be handled explicitly.

OmarManzoor replied:
The reason I popped it is because sample_weight is also explicitly being passed to the _log_reg_scoring_path method. If we pass sample_weight directly and then also pass it through the score_params it just seems to be duplicated.

adrinjalali commented on sklearn/linear_model/tests/test_logistic.py:
```diff
@@ -1984,3 +1985,33 @@ def test_warning_on_penalty_string_none():
         lr.fit(iris.data, target)
+
+
+def test_lr_cv_scorer_does_not_receive_sample_weight_when_score_request_is_not_set():
```
I quite like these two tests, but the tests you had in the common tests were also in the right direction. I don't think the tests here cover what we need to test:

In common tests we actually test if metadata is passed when it should, to the underlying objects, and here we don't do that. I think it'd be nice to have specific tests for LRCV in the common tests using the machinary there, even if they're not a part of the _common_ tests as the other ones.

OmarManzoor replied:
The problem with the ones in the common tests was that they were not actually working properly for scorers. The score method does not raise an exception in the inner implementation of MethodMetadataRequest, if we don't set the score request.

adrinjalali replied:
Can you please write a minimal test for it? It should raise, there shouldn't be a difference between `fit` and `score` in that regard.

OmarManzoor replied:
Could you kindly provide an example of a scorer with LogisticRegressionCV that would raise UnsetMetadataPassedError when sample_weight is passed but the score request is not set?

OmarManzoor replied:
This commit contains the precise part that is related to setting the unrequested params https://github.com/scikit-learn/scikit-learn/pull/24498/commits/f221ee9a0bf7681899851b73e9104eeba5ca5f75

adrinjalali replied:
generally it's not the _consumer_ which raises the error, rather the _router_ does that when `process_routing` is called. So the only thing we need to test is for the scorer to have the right request values.

adrinjalali replied:
yes, pretty much, very similar to how we do it in consumer estimators there.

OmarManzoor replied:
@adrinjalali The tests in CircleCI seem to be failing.

adrinjalali replied:
Seems to have something to do with `set_output` (I might be wrong). I'm checking. cc @thomasjpfan

OmarManzoor replied:
@adrinjalali The checks still seem to be failing but could you kindly check the test that I added in the common tests?

adrinjalali reviewed (commented):
This is great work, @OmarManzoor, thank you!

OmarManzoor committed:
Fix doc for fit_params

OmarManzoor committed:
Correct the passing of score params to _log_reg_scoring_path

OmarManzoor committed:
Add a test to test if metadata is routed correctly to scorer

OmarManzoor committed:
Inherit consuming scorer from base scorer

OmarManzoor committed:
Revert consuming scorer to check if tests pass

OmarManzoor committed:
Merge branch 'sample-props' into logistic_regression_cv_routing

OmarManzoor committed:
Add the ConsumingScorer and test back

adrinjalali commented:
Merging the latest `sample-props` branch should fix the CI issues: https://github.com/scikit-learn/scikit-learn/pull/24854

OmarManzoor committed:
Merge branch 'sample-props' into logistic_regression_cv_routing

OmarManzoor committed:
Add a consuming splitter and add a test to check routing for cv splitters

OmarManzoor commented:
@adrinjalali Could you kindly have a look at the updates?

jnothman commented on sklearn/linear_model/tests/test_logistic.py:
```diff
@@ -1984,3 +1985,33 @@ def test_warning_on_penalty_string_none():
+    assert not routed_params.scorer.score
+
+
+def test_lr_cv_scorer_receives_sample_weight_when_score_request_is_set():
```
We don't actually test that the weights are received or have an effect. Perhaps we should check that the results are different from when `sample_weight` is not requested.

OmarManzoor replied:
Thank you for the suggestion. I just wanted to mention that there seems to be an issue if I try to call fit on LogisticRegressionCV when requesting sample weight for the underlying scorer. The problem occurs at this point
https://github.com/scikit-learn/scikit-learn/pull/24498/files#diff-66422c7ed307653c80e5d68ed353d44d8fbb02319c428e140daff082eaf9d6cbR742. Over here given the sample weight has a length of 10 whereas y_test has a length of 2 and this causes an inconsistent number of samples error.

jnothman replied:
Please make sure to index a subset of weights for the train/test set like we do for X and y... Not that I've checked the code now

OmarManzoor replied:
@jnothman I updated the sample weight to only index the test set at the required place and also updated the test. Could you kindly have a look?

jnothman reviewed (approved):
Otherwise LGTM, thanks!

OmarManzoor committed:
Extract the test indices for sample weight when passing to the scorer and update the test

OmarManzoor committed:
Fix test

OmarManzoor committed:
Fix test

OmarManzoor committed:
Fix test

OmarManzoor committed:
Merge remote-tracking branch 'upstream/sample-props' into logistic_regression_cv_routing

adrinjalali commented on sklearn/linear_model/_logistic.py:
```diff
@@ -1751,6 +1760,9 @@ def fit(self, X, y, sample_weight=None):
 
+        **fit_params : dict
+            Parameters to pass to the underlying splitter and scorer.
+
```
the parameters for the public methods need a `.. versionadded:: 2.0`

adrinjalali reviewed (commented):
Thanks for the work so far @OmarManzoor

OmarManzoor committed:
Merge remote-tracking branch 'upstream/sample-props' into logistic_regression_cv_routing

OmarManzoor committed:
PR suggestions

adrinjalali commented:
@OmarManzoor we have switched to a feature flag system (introduced in https://github.com/scikit-learn/scikit-learn/pull/26103), which makes things much easier for implementing slep6 for meta-estimators.

The big slep6 PR is now merged into `main` (https://github.com/scikit-learn/scikit-learn/pull/24027), and the `sample-props` branch is now removed from the main repo.

Would you be up for converting this PR to the feature flag enabled form, and submit the PR to `main`?

Also happy to document/answer your questions on how to do the feature flag thingy.

OmarManzoor commented:
> @OmarManzoor we have switched to a feature flag system (introduced in #26103), which makes things much easier for implementing slep6 for meta-estimators.
> 
> The big slep6 PR is now merged into `main` (#24027), and the `sample-props` branch is now removed from the main repo.
> 
> Would you be up for converting this PR to the feature flag enabled form, and submit the PR to `main`?
> 
> Also happy to document/answer your questions on how to do the feature flag thingy.

Thank you for the suggestion. Sure, I would like to try this out as soon as I find time.
